src/ray/common/scheduling/resource_instance_set.cc

```diff
@@ -93,21 +123,130 @@ bool NodeResourceInstanceSet::operator==(const NodeResourceInstanceSet &other) c
+          std::make_pair(resource_id, data.value()));
     } else {
-      // Allocation failed. Restore partially allocated resources.
-      for (const auto &[resource_id, allocation] : allocations) {
-        Free(resource_id, allocation);
+      // Directly allocate the resources if the resource is not with a placement group
+      auto allocation = TryAllocate(resource_id, demand);
+      if (allocation) {
+        // Even if allocation failed we need to remember partial allocations to
+        // correctly free resources.
+        allocations[resource_id] = std::move(*allocation);
+      } else {
+        // Allocation failed. Restore partially allocated resources.
+        for (const auto &[resource_id, allocation] : allocations) {
+          Free(resource_id, allocation);
+        }
+        return std::nullopt;
       }
-      return std::nullopt;
     }
   }
 
+  // Handle the resource allocation for resources with placement group
+  for (const auto &[original_resource_id, resource_id_vector] : pg_resource_map) {
+    // Assuming exactly 1 placement group and at most 1 bundle index can be specified in
+    // the resource requirement for a single resource type
+    std::vector<FixedPoint> wildcard_allocation;
+    const ResourceID *wildcard_resource_id = nullptr;
+    const std::string *pg_id = nullptr;
+
+    // Allocate indexed resource
+    if (resource_id_vector.size() == 1) {
+      // The case where no bundle index is specified
+      // Iterate through the bundles with the same original resource and pg_id to find
+      // the first one with enough space
+      bool found = false;
+      wildcard_resource_id = &resource_id_vector[0].first;
+      pg_id = &resource_id_vector[0].second.group_id;
+
+      auto pg_index_resources = pg_indexed_resources_.find(original_resource_id);
+      if (pg_index_resources != pg_indexed_resources_.end()) {
+        auto index_resources = pg_index_resources->second.find(*pg_id);
+        if (index_resources != pg_index_resources->second.end()) {
+          for (ResourceID indexed_resource_id : index_resources->second) {
+            if (Has(indexed_resource_id)) {
+              auto allocation = TryAllocate(
+                  indexed_resource_id, resource_demands.Get(resource_id_vector[0].first));
+
+              if (allocation) {
+                // Found the allocation in a bundle
+                wildcard_allocation = *allocation;
+                allocations[indexed_resource_id] = std::move(*allocation);
+                found = true;
+                break;
+              }
+            }
+          }
+        }
+      }
+
+      if (!found) {
+        // No bundle can fit the required resources, allocation failed
+        for (const auto &[resource_id, allocation] : allocations) {
+          Free(resource_id, allocation);
+        }
+        return std::nullopt;
+      }
+    } else {
+      // The case where the bundle index is specified
```

Comment:
is there no case where the demands only include indexed resources, but not wildcard resources? I assume no?

Reply:
(feel like the logic assumes this to be correct. maybe add it to the docstring?)

Reply:
IIUC, here: https://github.com/ray-project/ray/blob/d80b22f704634df46445efde10b99f2492781892/src/ray/common/bundle_spec.cc#L199 it will always include the wildcard resource. 

At the same time, I'll add the assumption to the comment as well.